core/store/src/trie/mod.rs

```diff
@@ -658,18 +666,20 @@ impl Trie {
         }
     }
 
-    pub fn get_ref(
-        &self,
-        root: &CryptoHash,
-        key: &[u8],
-    ) -> Result<Option<(u32, CryptoHash)>, StorageError> {
-        let key = NibbleSlice::new(key);
-        self.lookup(root, key)
+    pub fn get_ref(&self, root: &CryptoHash, key: &[u8]) -> Result<Option<ValueRef>, StorageError> {
+        let is_delayed = is_delayed_receipt_key(key);
+        match &self.flat_state {
+            Some(flat_state) if !is_delayed => flat_state.get_ref(&key),
```

Comment:
Yeah, so this bit feels suspicious to me: what if `root` happens to be some historical, old root, for which  the data isn't actually present in the `flat_state`? 

Perhaps a `flat_state` should be tagged with a `root` this state is for, so the logic reads like this: 

```
pub fn get_ref(
        &self,
        root: &CryptoHash,
        key: &[u8],
    ) -> Result<Option<(u32, CryptoHash)>, StorageError> {
        // Fast path -- fetch from flat_state in one step
        if self.flat_state.root == root { return flat_state.get_ref(key) }
        // Slow path -- we don't have flat state or are querying historical data, 
        // fall-back to lookup from trie
        let key = NibbleSlice::new(key);
        self.lookup(root, key)
    }

```

Reply:
It is not supported yet, I will add TODO for it.

Reply:
Added TODO with clarification - we assume that flat state always contains data for the given root.